Readme.md

```diff
@@ -129,13 +129,276 @@ with keys corresponding to the options described in [CLI Arguments](#cliargs):
+  previously-executed parsers.
+
+* The `context` object represents the options, settings, and other data used by the log
+  processors. The `options` object used to initialize Pino, is merged into the `context`
+  object, making those settings available to overy log processor.
+
+  The following properties are always available as well:
+  - `EOL`: the actual end-of-line characters, as specified by the `crlf` option
+  - `IDENT`: the default indentation string
+  - `translateFormat`: the time formatting string, as specified by the `translateTime`
+    option
+  - `colorizer`: the selected colorizer function that accepts a level value and
+    returns a colorized string
+  - `prettified`: an object that caches prettified text, whose properties are typically
+    set by a parser and consumed by a builder
+
+* The `state` object has a `stop` method, which is used to abort the log processing
+  process for the current entry immediately after the current parser returns.
+
+The return value of the parse function is passed to the next parse function in the
+sequence. The value returned from the last parse function is passed to the build
+function.
+
+For most parsers, the return value should be a log entry object that will be used by
+the builders to construct the output line.
+
+In some cases, whether for performance or other reasons, it makes sense to return the
+result of a parser directly without performing any additional processing. The process
+can be short-circuited by calling the `stop` method of the `state` argument. No other
+parsers will be used, and the build stage will be skipped as well. The value returned
+by the parser that calls the `stop` function will be returned as the final formatted
+output.
+
+#### building
+
+The `build` function is a simple functions that prepares the final formatted line.
+It receives two parameters: `lineParts` and `context`.
+* The `lineParts` array contains the ordered list of strings that will eventually be joined
+  after all the build functions have been executed.
+* The `context` object is the same `context` object passed to the parsers, with one
+  exception: the final output produced by the parsers is added to the object as the `log`
+  property.
+
+  Also, when the built-in prettifying log processors are used, a number of prettified
+  strings are added to the `context.prettified` property: `prettifiedLevel`,
+  `prettifiedMessage`, `prettifiedMetadata`, and `prettifiedTime`.
+
+The builder function does not return a value. To change the output, modify the
+`lineParts` array.
+
+Note that if a parser short-circuits the parsing process, the builders will not be
+executed at all.
+
+#### examples
+
+The following are a number of examples of parsers, builders, and combined log processors:
+
+* **parser** - Perform a transformation on the input message:
+
+  ```js
+  function (input) {
+    input.msg = input.msg.toLowerCase()
```

Comment:
I'm torn. I don't like functions that have side effects. But the immediately obvious alternative incurs even more overhead. But maybe, and this is just a thought for discussion, the signature could be `function parser ({logObject, context, state, collector}) {}`. `logObject` would be a frozen instance of the originally parsed log line and the `collector` could be modified and passed along to ultimately be compiled into the output string.

@mcollina @davidmarkclements what are your opinions?

Reply:
I have to say that I share your concern. While I was working on this, I toyed with using pure functions that return an object like this (pseudo-code):

```js
const result = {
  output: { ...input, modifiedProperty, newProperty },
  done: false || true, // the previous incarnation of state.stopped
}
```

The problem with this approach is that it would almost certainly incur a significant hit to performance associated with the rapid creation and destruction of so many objects and primitives. Also, this approach would only create a shallow clone. A deep clone would hurt performance even more.

So that leaves us with unavoidable side-effects. A mitigating factor though is that the pipeline goes in one direction and we don't use complicating features such as events, rendering the side-effects relatively safe.

One of the problems with my approach that all of you have (thankfully!) pointed out in one way or another is that it's not very clear where to apply the side-effects: modify the `input` object, add values to the `context.prettified` object, or modify the `context` object directly.

The `collector` might be able to solve that problem, but it if it doesn't preserve the original structure of the log object, it would force us to rethink the way the builders work.

Also, if we used the the `collector`, it should be the second argument, not the last. `context` and `state` aren't always necessary, the latter being especially rare, but the collector would be used in almost every parser.